95.md

```diff
@@ -0,0 +1,108 @@
+The base64 data will be used on the `.content` field of one or more `kind:1064` events to be uploaded to a relay.
+
+The chunk events may be sent on separate websockets connections for parallel uploads.
+
+## Event
+
+An event of `kind:1064` represents a file chunk. A client may choose to split a file
+into multiple chunks or use a single event.
+
+The event `.pubkey` is NOT the uploader's one but a randomly generated pubkey. All chunks
+use the same pubkey. The pubkey MUST NOT be reused on events other than the
+`kind:1064` event set of a single file.
+**The pubkey is what is used to uniquely identify the file made of a set of chunk events**
+
+The event `.content` is the base64-encoded file data.
+
+It has a `c` tag. Its first falue is the chunk number while the second one is the total number of chunks.
+The client can use the `c` tag to ask for missing chunks when resuming a download.
+
+Example:
+
+```js
+{
+  "id": "<id>",
+  "pubkey": "<same-random-pubkey-for-all-event-chunks>"
+  "kind": 1064,
+  "tags": [
+    ["c", "5", "20"] // required; this is the 5th chunk of a total of 20
+  ],
+  "content": "<base64-chunk>",
+  "created_at": <timestamp>,
+  "sig": "<signature>"
+}
+```
+
+## Metadata
+
+Metadata should be added using [NIP-94](94.md) or other metadata-related NIP.
+
+## File Relay
+
+Client should upload to user's "file relays", which use the [NIP-65](65.md) `f` flag.
+When downloading a file uploaded with this NIP, it should search on the uploader's "file relays".
+
+**Relays must NOT honor `kind:5` deletion events referencing file chunk events.** Deletion
+is expected to be automatic when no uploader is registered (see [Pre-Upload](#pre-upload) section
+to learn how to register an uploader). If an user wants to make sure a file won't be deleted, it
+should become an uploader of that file.
+
+## Pre-Upload
+
+Before uploading `kind:1064` events of a specifc file,
+user MUST publish a single `kind:1065` "Uploader" event,
+which is an event authored by the **user's main pubkey**,
+with empty `.content` and an `f` tag referencing the `kind:1064`
+event set (the `kind:1064` event(s) `.pubkey`).
+
+Its purpose is to register the user as an uploader of the file on a specific relay and to request permission
+for uploading the file chunks.
+
+When receiving the `kind:1065` event, relay may answer with one of the following messages and prefixes:
+
+- `["OK", "<kind:1065-event-id>", false, "auth-required: ..."]`: Authentication is required to register an uploader;
+- `["OK", "<kind:1065-event-id>", true, "uploaded: ..."]`: The corresponding `kind:1064` file chunks are already uploaded, trying to re-upload them will fail;
+- `["OK", "<kind:1065-event-id>", true, "upload: Missing chunks 1, 2, 7, 10"]`: File isn't uploaded yet or incomplete, user is allowed to upload it on this ws connection;
+
+Trying to send a `kind:1064` event before a `kind:1065` one should fail.
```

Comment:
I don't like these custom behaviors for relays.

Reply:
The reason for this is that a file relay can't let a client send a big event (a file chunk) just to later reply that the client/user had no authorization to do it. It wastes relay resources so I imagine file relay will start with a capped max ws message size for every new ws connection until it sees a `kind:1065` event "asking" for authorization to upload a `kind:1064` event.